man/man1/cwb_setcuts.1
.\" Man page generated from reStructuredText.
.
.TH "CWB_SETCUTS" "1" "Jan 14, 2019" "" "coherent WaveBurst"
.SH NAME
cwb_setcuts \- cWB Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.nf

.fi
.sp
.nf

\fBUsage\fP

This command applies thresholds on tree leaves to select events with certain characteristics defined by the user.
We remind that for each merge label we have two files type: the wave and the mdc/live, with different structure but some leafs are in common.
So when choosing selection cuts, you should distinguish between the cuts that should be applied only to wave tree from the ones that will be
applied also to mdc/live. One example of the former cut is a selection on rho or cc, on the contray a selection on the run is an example of the latter.

\fBSyntax\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_setcuts mlabel/list wcuts clabel jcuts(opt) mcuts(opt) lcuts(opt)\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBmlabel/lists\fP same label used by cwb_merge, if the user writes lists it list the available labels
.IP \(bu 2
\fBclabel\fP label that will be added to the new merged files names (M1 \-> M1.C_clabel)
.IP \(bu 2
\fBwcuts\fP selections for only wave tree must contains only leaf of waveburst tree (not the leaf run) the syntax is the the same as the one used for the selection TTree
.IP \(bu 2
\fBjcuts\fP job selection for wave & mdc trees must contains only the leaf run the syntax is the the same as the one used for the selection TTree
.IP \(bu 2
\fBmcuts\fP selections for only mdc tree must contains only leaf of mdc tree (not the leaf run) the syntax is the the same as the one used for the selection TTree
.IP \(bu 2
\fBlcuts\fP selections of lag/slag : is applied only to wave and livetime tree must contains only lag/slag leafs the syntax is the the same as the one used for the selection TTree
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.nf

\fBAlternative Syntax\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_setcuts mlabel/list options\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBmlabel/lists\fP same label used by cwb_merge
.IP \(bu 2
\fBoptions\fP options (options must be declared inside \(aq...\(aq) : see previous definitions
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_setcuts mlabel/list wcuts clabel jcuts(opt) mcuts(opt) lcuts(opt)\fP
.IP \(bu 2
\fB\-\-label clabel\fP default=""
.IP \(bu 2
\fB\-\-tcuts tcuts\fP default="" : tcuts is the name of TCut selection defined in user_pparameters.C, it is added to wcuts
.IP \(bu 2
\fB\-\-wcuts wcuts\fP default=""
.IP \(bu 2
\fB\-\-jcuts jcuts\fP default=""
.IP \(bu 2
\fB\-\-mcuts mcuts\fP default=""
.IP \(bu 2
\fB\-\-lcuts lcuts\fP default=""
.IP \(bu 2
\fB\-\-unique true/false\fP select from the root file a unique reconstructed event for each injected signal.
.INDENT 2.0
.INDENT 3.5
The signal is selected from the reconstructed events which have the same injected gps time, the selected one has the maximum rho.
When this option is used the label ".C_U" will be added to the new merged files names. (default=false)
A new leaf is added to the waveburst tree : fsize=number of fragments found in the \fI\%iwindow\fP
NOTE : this option can be applied only when simulation>0.
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
not use \fB!expression\fP but \fB!(expression)\fP
EX : ot use !veto_cat3_H1 but !(veto_cat3_H1)
.TP
use \fB!expression\fP instead of \fB"expression"\fP
EX : use !veto_cat3_H1 instead of "!(veto_cat3_H1)"
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.nf

\fBExamples\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
* cwb_setcuts M1 \(aqnetcc[1]>0.8\(aq  cc1_gt_0d8_run_lt_100 \(aqrun<100\(aq
  \- apply the cut ((netcc[1]>0.8)&&(run<100)) to wave file
  \- apply the cut (run<100) to the live file

  if the input file is with name: merge/wave_S6A_BKG_L1H1_run1.M1.root
  the output file assumes the name: merge/wave_S6A_BKG_L1H1_run1.M1.C_cc1_gt_0d8_run_lt_100.root
  and the same for live file.

  Alternative Syntax :
  cwb_setcuts M1 \(aq\-\-wcuts netcc[1]>0.8  \-\-label cc1_gt_0d8_run_lt_100 \-\-jcuts run<100\(aq

* cwb_setcuts M1.C_rho0_gt6.V_cat2LH_hvetoLH_cat3LH  \(aq!(veto_cat3_H1) && !(veto_cat3_L1)\(aq nohwinj

* cwb_setcuts M1 \(aqnetcc[2]>0.45\(aq  cc2_gt_0d45_lag1_eq_0 \(aq\(aq \(aq\(aq \(aqlag[1]==0\(aq
  \- apply the cut ((netcc[2]>0.45)&&(lag[1]==0)) to wave file
  \- apply the cut (lag[1]==0) to the live file

  Alternative Syntax :
  cwb_setcuts M1 \(aq\-\-wcuts netcc[2]>0.45 \-\-label cc2_gt_0d45_lag1_eq_0 \-\-lcuts lag[1]==0\(aq

* cwb_setcuts M1 \(aq\-\-unique true\(aq
  \- apply unique selection of detected events

  if the input file is with name: merge/wave_S6A_BKG_L1H1_run1.M1.root
  the output file assumes the name: merge/wave_S6A_BKG_L1H1_run1.M1.C_U.root

* cwb_setcuts M1 \(aq\-\-unique true \-\-wcuts netcc[2]>0.45  \-\-label cc2_gt_0d45\(aq
  \- apply unique selection of detected events and tree cuts

  if the input file is with name: merge/wave_S6A_BKG_L1H1_run1.M1.root
  the output file assumes the name: merge/wave_S6A_BKG_L1H1_run1.M1.C_U.C_cc2_gt_0d45.root

* cwb_setcuts M1 \(aq\-\-tcuts MyTCut \-\-wcuts netcc[2]>0.45 \-\-label cc2_gt_0d45_mytcut\(aq
  \- select events with (MyTCut.GetTitle()) && (netcc[2]>0.45)

  The MyTCut must be defined in user_pparameters.C
  Ex : TCut MyTCut(\(aqMyTCut\(aq,\(aqrho[0]>6\(aq); // MyTCut.GetTitle() is \(aqrho[0]>6\(aq
  or it can be defined in a file.hh and included in user_pparameters.C inside the brackets {}
  Ex : #include "file.hh"
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SH AUTHOR
Gabriele Vedovato
.SH COPYRIGHT
2018, cWB Team
.\" Generated by docutils manpage writer.
.


tools/cwb/macros/cwb_setcuts.C
/*
# Copyright (C) 2019 Gabriele Vedovato
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/


// apply selection on leaves to merged root file : used by the cwb_setcuts command

{

  CWB::Toolbox TB;

  int estat;
  Long_t id,size,flags,mt;
  char cmd[1024];

  	  cwb_merge_label     = TString(gSystem->Getenv("CWB_MERGE_LABEL"));
  TString cwb_jcuts_tree      = TString(gSystem->Getenv("CWB_JCUTS_TREE"));
  TString cwb_wcuts_tree      = TString(gSystem->Getenv("CWB_WCUTS_TREE"));
  TString cwb_cuts_label      = TString(gSystem->Getenv("CWB_CUTS_LABEL"));
  TString cwb_mcuts_tree      = TString(gSystem->Getenv("CWB_MCUTS_TREE"));
  TString cwb_lcuts_tree      = TString(gSystem->Getenv("CWB_LCUTS_TREE"));
  TString cwb_setcuts_options = TString(gSystem->Getenv("CWB_SETCUTS_OPTIONS"));

  TString mdir = merge_dir;

  // check if cwb_setcuts_options contains '--' than cwb_setcuts_options
  // is used to extract all setcuts parameters
  TString cwb_tcuts_tree = "";
  TString cwb_unique_evt = "false";
  if(cwb_setcuts_options.Contains("--")) {
    TString option="";
    // get the cwb_tcuts_tree
    cwb_tcuts_tree = TB.getParameter(cwb_setcuts_options,"--tcuts");
    // get the cwb_wcuts_tree
    cwb_wcuts_tree = TB.getParameter(cwb_setcuts_options,"--wcuts");
    // get the cwb_jcuts_tree
    cwb_jcuts_tree = TB.getParameter(cwb_setcuts_options,"--jcuts");
    // get the cwb_mcuts_tree
    cwb_mcuts_tree = TB.getParameter(cwb_setcuts_options,"--mcuts");
    // get the cwb_lcuts_tree
    cwb_lcuts_tree = TB.getParameter(cwb_setcuts_options,"--lcuts");
    // get the cwb_cuts_label
    cwb_cuts_label = TB.getParameter(cwb_setcuts_options,"--label");
    // get the cwb_unique_evt
    cwb_unique_evt = TB.getParameter(cwb_setcuts_options,"--unique");
  }
  // if cwb_tcuts_tree!="" than cwb_tcuts_tree is added to cwb_wcuts_tree
  if(cwb_tcuts_tree!="") {
     // check if tcut is defined
     TGlobal *global=(TGlobal*)gROOT->GetListOfGlobals()->FindObject(cwb_tcuts_tree.Data());
     if(global==NULL) {
        cout << "cwb_setcuts.C : Error - tcut " << cwb_tcuts_tree << " is not defined " << endl; 
        cout << "                must be included in the user_pparameters.C file" << endl;
        gSystem->Exit(1);
     }
     TCut *tcut = (TCut*)global->GetAddress();
     if(TString(tcut->GetName())!=cwb_tcuts_tree) {
        cout << "cwb_setcuts.C : Error - tcut " << cwb_tcuts_tree << " not correspond to global Tcut name" << endl; 
        cout << "                must be included in the user_pparameters.C file" << endl;
        gSystem->Exit(1);
     }
     cwb_tcuts_tree = tcut->GetTitle();	// extract TCut 
     cout << "cwb_tcuts_tree : " << cwb_tcuts_tree << endl;
     if(cwb_wcuts_tree=="") {
        cwb_wcuts_tree = cwb_tcuts_tree;
     } else {
       TString cwb_wcuts = cwb_tcuts_tree;
       cwb_wcuts = TString("(")+cwb_wcuts_tree+")&&("+cwb_tcuts_tree+")";
       cwb_wcuts_tree = cwb_wcuts;
     }
     cout << "cwb_wcuts_tree : " << cwb_wcuts_tree << endl;
  }
  // check if cwb_unique_evt can be applied 
  if((simulation==0)&&(cwb_unique_evt=="true")) {
    cout << "cwb_setcuts.C : Error - unique option can be applied only to simulated data" << endl; 
    gSystem->Exit(1);
  }
  // check if cwb_cuts_label is defined
  if((cwb_wcuts_tree!="")&&(cwb_cuts_label=="")) {
    cout << "cwb_setcuts.C : Error - cuts label not defined" << endl; 
    gSystem->Exit(1);
  }
  // check if cwb_cuts_label contains '.'
  if(cwb_cuts_label.Contains(".")) {
    cout << "cwb_setcuts.C : Error - cwb_cuts_label " << cwb_cuts_label 
         << " can not contains '.'" << endl << endl;
    gSystem->Exit(1);
  }
  // check if cwb_wcuts_tree contains run leaf
  if(cwb_wcuts_tree.Contains("run")) {
    cout << "cwb_setcuts.C : Error - cwb_wcuts_tree " << cwb_jcuts_tree 
         << " can not contains leaf : run" << endl << endl;
    gSystem->Exit(1);
  }
  // check if cwb_wcuts_tree contains slag leaf
  if(cwb_wcuts_tree.Contains("slag")) {
    cout << "cwb_setcuts.C : Error - cwb_wcuts_tree " << cwb_jcuts_tree 
         << " can not contains leaf : slag" << endl << endl;
    gSystem->Exit(1);
  }
  // check if cwb_wcuts_tree contains lag leaf
  if(cwb_wcuts_tree.Contains("lag")) {
    cout << "cwb_setcuts.C : Error - cwb_wcuts_tree " << cwb_jcuts_tree 
         << " can not contains leaf : lag" << endl << endl;
    gSystem->Exit(1);
  }
  if(cwb_wcuts_tree=="") cwb_wcuts_tree="run>0";
  if(cwb_jcuts_tree=="") cwb_jcuts_tree="run>0";

  // create wcuts
  TString cwb_wcuts = cwb_wcuts_tree;
  cwb_wcuts = TString("(")+cwb_wcuts_tree+")&&("+cwb_jcuts_tree+")";
  if(cwb_lcuts_tree!="") cwb_wcuts = TString("(")+cwb_wcuts+")&&("+cwb_lcuts_tree+")";

  // create mcuts
  TString cwb_mcuts = cwb_mcuts_tree;
  if(cwb_mcuts!="") 
    cwb_mcuts = TString("(")+cwb_mcuts_tree+")&&("+cwb_jcuts_tree+")";
  else
    cwb_mcuts = cwb_jcuts_tree;

  // create input wave root cuts file name
  char iwfname[1024];  
  sprintf(iwfname,"wave_%s.%s.root",data_label,cwb_merge_label.Data());

  // create output wave root cuts file name
  TString uwfname = mdir+"/"+iwfname;
  uwfname.ReplaceAll(".root",TString(".C_U.root"));

  // if cwb_unique_evt=true then create a wave root file with unique
  // reconstructed events
  if(cwb_unique_evt=="true") {
    TB.setUniqueEvents(mdir+"/"+iwfname, uwfname, nIFO, pp_irho);
    // create symbolic link of mdc file
    char imfname[1024];  
    sprintf(imfname,"mdc_%s.%s.root",data_label,cwb_merge_label.Data());
    TString omfname = uwfname;
    omfname.ReplaceAll("wave_","mdc_");
    omfname.Remove(0,omfname.Last('/')+1);	// strip path
    cout << omfname << endl;
    estat = gSystem->GetPathInfo(mdir+"/"+imfname,&id,&size,&flags,&mt);
    if(estat==0) {
      sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),imfname,omfname.Data());
      cout << cmd << endl;
      gSystem->Exec(cmd);
    }
    // create symbolic link of merge list file
    char ilstfname[1024];  
    sprintf(ilstfname,"merge_%s.%s.lst",data_label,cwb_merge_label.Data());
    TString olstfname = uwfname;
    olstfname.ReplaceAll("wave_","merge_");
    olstfname.ReplaceAll(".root",".lst");
    olstfname.Remove(0,olstfname.Last('/')+1);	// strip path
    cout << olstfname << endl;
    estat = gSystem->GetPathInfo(mdir+"/"+ilstfname,&id,&size,&flags,&mt);
    if (estat==0) {
      sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilstfname,olstfname.Data());
      cout << cmd << endl;
      gSystem->Exec(cmd);
    }
  }
  if((cwb_wcuts_tree="")&&(cwb_cuts_label=="")) exit(0);

  if(cwb_unique_evt=="true") {
    sprintf(iwfname,"wave_%s.%s.C_U.root",data_label,cwb_merge_label.Data());
    cwb_merge_label+=".C_U";
  }

  // create output wave root cuts file name
  TString owfname = mdir+"/"+iwfname;
  owfname.ReplaceAll(".root",TString(".C_")+cwb_cuts_label+".root");

  // create a merge*.lst file name & run selection cuts
  vector<int> jobList;
  char ilstfname[1024];  
  sprintf(ilstfname,"merge_%s.%s.lst",data_label,cwb_merge_label.Data());
  TString olstfname = owfname;
  olstfname.ReplaceAll("wave_","merge_");
  olstfname.ReplaceAll(".root",".lst");
  olstfname.Remove(0,olstfname.Last('/')+1);	// strip path
  cout << olstfname << endl;
  if(cwb_jcuts_tree=="") {	// no jobs cuts -> create a symbolic link
    estat = gSystem->GetPathInfo(mdir+"/"+ilstfname,&id,&size,&flags,&mt);
    if (estat==0) {
      sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilstfname,olstfname.Data());
      cout << cmd << endl;
      gSystem->Exec(cmd);
    }
  } else {
    // build list of selected jobs
    vector<TString> jobFileList;
    vector<int> jobList = CWB::Toolbox::getMergeJobList(mdir+"/"+ilstfname,jobFileList);
    cout << "Number of jobs in the merged lists : " << jobList.size() << endl;
    // create a dummy tree, fill with merged jobs, apply cuts, get selected jobs
    TTree jtree("jtree","jtree");
    int run; jtree.Branch("run",&run,"run/I");
    TString *jfname = new TString();
    jtree.Branch("jfname", &jfname);
    // fill with merged jobs
    for(int i=0;i<jobList.size();i++) {run=jobList[i]; *jfname=jobFileList[i]; jtree.Fill();}
    // check cuts
    TTreeFormula formula("trCuts", cwb_jcuts_tree.Data(), &jtree);
    int err = formula.Compile(cwb_jcuts_tree.Data());
    if(err) {
      cout << "cwb_setcuts.C - wrong input cuts " << cwb_jcuts_tree << endl << endl;
      gSystem->Exit(1);
    }
    // get selected jobs 
    jtree.Draw("Entry$",cwb_jcuts_tree.Data(),"goff");
    int jsel = jtree.GetSelectedRows();
    double* entry = jtree.GetV1();			// get selected jobs
    jobList.clear();
    jobFileList.clear();
    for(int i=0;i<jsel;i++) {
      jtree.GetEntry(entry[i]);
      //cout << i << " " << run << " " << jfname->Data() << endl;
      jobList.push_back(run);				// list of selected jobs
      jobFileList.push_back(jfname->Data());		// list of selected file jobs
    }
    CWB::Toolbox::dumpFileList(jobFileList,mdir+"/"+olstfname);
  }

  // apply wcuts to wave file
  int nsel = CWB::Toolbox::setCuts(iwfname,mdir,mdir,"waveburst",cwb_wcuts,cwb_cuts_label);
  if(nsel<=0) {
    cout << "cwb_setcuts.C : Error - Number of selected waveburst entries = " << nsel << endl << endl;
    gSystem->Exit(1);
  }
  cout << "cwb_setcuts.C : Number of selected waveburst entries = " << nsel << endl;

  // apply mcuts to mdc file
  if(simulation) {
    char imfname[1024];  
    sprintf(imfname,"mdc_%s.%s.root",data_label,cwb_merge_label.Data());
    int nsel = CWB::Toolbox::setCuts(imfname,mdir,mdir,"mdc",cwb_mcuts,cwb_cuts_label);
    cout << "cwb_setcuts.C : Number of selected mdc entries = " << nsel << endl << endl;
    if(nsel<0) { // create a symbolic link to mdc*.root file name
      TString omfname = owfname;
      omfname.ReplaceAll("wave_","mdc_");
      omfname.Remove(0,omfname.Last('/')+1);	// strip path
      cout << omfname << endl;
      estat = gSystem->GetPathInfo(mdir+"/"+imfname,&id,&size,&flags,&mt);
      if(estat==0) {
        sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),imfname,omfname.Data());
        cout << cmd << endl;
        gSystem->Exec(cmd);
      }
    }
  }

  // apply cuts to live file
  if(!simulation) {
    char ilfname[1024];  
    sprintf(ilfname,"live_%s.%s.root",data_label,cwb_merge_label.Data());
    TString cwb_live_cuts = cwb_jcuts_tree;
    if(cwb_lcuts_tree!="") cwb_live_cuts = TString("(")+cwb_live_cuts+")&&("+cwb_lcuts_tree+")";
    int nsel = -1;
    if(cwb_live_cuts!="run>0") {
      nsel = CWB::Toolbox::setCuts(ilfname,mdir,mdir,"liveTime",cwb_live_cuts,cwb_cuts_label);
      cout << "cwb_setcuts.C : Number of selected live entries = " << nsel << endl;
    }
    if(nsel<0) { // create a symbolic link to live*.root file name
      TString olfname = owfname;
      olfname.ReplaceAll("wave_","live_");
      olfname.Remove(0,olfname.Last('/')+1);	// strip path
      cout << olfname << endl;
      estat = gSystem->GetPathInfo(mdir+"/"+ilfname,&id,&size,&flags,&mt);
      if(estat==0) {
        sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilfname,olfname.Data());
        cout << cmd << endl;
        gSystem->Exec(cmd);
      }
    }
  }

  exit(0);
}
